### PR TITLE
🧹 `Journal`: `Entries` store their `Keywords`

### PR DESCRIPTION
**Pull request by zspencer**
- https://github.com/zinc-collective/convene/issues/1662
- https://github.com/zinc-collective/convene/issues/1566

This gets us a little bit closer to being able to browse `Entry` by `Keyword`s, since each `Entry` will know it's keywords

### CODE REVIEW

File: app/furniture/journal/keyword.rb [hunk 1]
```diff
@@ -7,9 +7,11 @@ class Keyword < ApplicationRecord
     belongs_to :journal, inverse_of: :keywords
 
     def self.extract_and_create_from!(body)
-      body.scan(/#(\w+)/)&.flatten&.each do |keyword|
-        next if where(":aliases = ANY (aliases)", aliases: keyword)
-          .or(where(canonical_keyword: keyword)).exists?
+      body.scan(/#(\w+)/)&.flatten&.map do |keyword|
+        existing_keyword = where(":aliases = ANY (aliases)", aliases: keyword)
+          .or(where(canonical_keyword: keyword)).first
```
Comment (anaulin) [inline app/furniture/journal/keyword.rb:12]: I realize this query is from before your PR, but reading it over right now it struck me that it feels like it should be a nice little scope, which would help tighten up this entire loop a bit.

File: db/migrate/20230713230334_journal_add_keywords_to_entries.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class JournalAddKeywordsToEntries < ActiveRecord::Migration[7.0]
+  def change
+    add_column :journal_entries, :keywords, :string, array: true
```
Comment (anaulin) [inline db/migrate/20230713230334_journal_add_keywords_to_entries.rb:3]: I'm surprised these are not tied in any way to the ids of the Journal keywords -- I would have expected you'd want that to show "all entries with keyword X in Journal".

**Review by anaulin — APPROVED** [review]
Having keywords at both the Journal and the Entry level feels a little bit overkill to me, but I'm sure you have some larger plan for what that's necessary.

**Review by daltonrpruitt — APPROVED** [review]
I spent too long trying to figure out where the keywords were shown in the UI....
I should read more deeply. 😬

File: app/furniture/journal/keyword.rb (hunk 1 above)
Comment (zspencer) [reply app/furniture/journal/keyword.rb:12]: Agree; I would love for this to be a scope!

File: db/migrate/20230713230334_journal_add_keywords_to_entries.rb (hunk 2 above)
Comment (zspencer) [reply db/migrate/20230713230334_journal_add_keywords_to_entries.rb:3]: You know, I was thinking a join table would be "too much" but I suppose it's not?

The way I was planning to find the `Entries` was indexing this keywords field and using the scope you suggested above; but I do worry that that would make the relationy bits a bit less railsy

**Comment by zspencer** [comment]
> Having keywords at both the Journal and the Entry level feels a little bit overkill to me, but I'm sure you have some larger plan for what that's necessary.

I was thinking that an `Entry` >=< `Keyword` relationship is a many-to-many, and the `Journal` =< `Keyword` is one-to-many. 

> I spent too long trying to figure out where the keywords were shown in the UI.... 
> I should read more deeply. 😬

I am trying to take very very tiny steps; instead of just YOLO-Shipping the whole thing lol. It does make it a bit harder to figure out the trajectory tho.

File: app/furniture/journal/keyword.rb [hunk 3]
```diff
@@ -6,10 +6,16 @@ class Keyword < ApplicationRecord
     validates :canonical_keyword, presence: true, uniqueness: {case_sensitive: false, scope: :journal_id}
     belongs_to :journal, inverse_of: :keywords
 
+    scope(:search, lambda do |*keywords|
+      where("lower(aliases::text)::text[] && ARRAY[?]::text[]", keywords.map(&:downcase))
+          .or(where("lower(canonical_keyword) IN (?)", keywords.map(&:downcase)))
```
Comment (zspencer) [inline app/furniture/journal/keyword.rb:11]: There is probably a way less awkward way of doing this; but this is what I was able to come up with 🙃

**Review by zspencer — COMMENTED** [review]
OK, I've extracted the scope; and I am wanting to see how far I can get before sprouting the join table


The reason I'm avoiding the join table is because I work in another code base with hundreds of thousands of "taggings" in a join table; and the performance characteristics are pretty bad; even with fully indexing the taggings table across the tag_id and entry_id in both directions!

My gut says that storing the tags an Entry uses and indexing *that* will be far more performant because using the canonical keyword as a kind of "natural join" when traversing from the keyword to the entries; and from the entries to the keywords.

But I'm willing to drop that concern.

**Pull request merged**